### DISCUSSION TITLE
Potential Crash on iOS: std::runtime_error

### DISCUSSION THREAD

**Discussion by georgbachmann** [maplibre/maplibre-native#841] [Q&A]
**Describe the bug**
I have no idea if my problem is related to mapbox-libre, but the one big thing I changed from my previous release to the latest, is switching from the official mapbox version to this mapbox-libre fork.

I am using the latest version here and I am seeing a super strange crash in Firebase with no real hint, except `Fatal Exception: std::runtime_error` and I noticed that all my crashed devices had very little RAM left over.
I have not been able to reproduce this though myself... this is why I am vaguely asking if that could be related to maybe https://github.com/maplibre/maplibre-gl-native/pull/632 ?!? as it seems that Mapbox is throwing exceptions like this and as this issue here also is something about memory-access?!?

Is your fix @tdcosta100 already merged, right? And is it part of one of the latest iOS release as well?

Sorry for my vague question, I just don't have more infos available currently. As firebase might add more, I'll update here. Also Xcode is not reporting any issues at all. So there will hopefully be more info once I get those.

**Platform information:**
 - OS: iOS
 - iOS Version: from 15 - 16 all versions
 - Mapbox-Version 5.13.0

**Comment by louwers** [comment]
There has been a new release since that was merged (5.13.0). But a `std::runtime_error` could be anything. Do you see any exception message?

More information is needed to look into this.

**Comment by georgbachmann** [comment]
@louwers thanks for your help. I know my bugreport is very very vague... I don't have a lot more information on that currently. Still all my stack-traces are one single line with this `std::runtime_error` which yes, could be everything. But I know from logs that users are at least interacting with the map when it happens and the move to MapboxLibre was the major change in the latest update... so yes, it's just a wild guess... but my gut feeling...
About that #632 PR... How can I check if it made it into that latest iOS build? I am asking cause that handles an exception that is thrown when getting features to render (if I understand the code correctly... my c++ skills are not that good...?!?)
Also, do you know how often they are made and when the next one is due?
Cheers

**Reply by louwers** [reply]
There is no set release schedule at this point. You may want to try `#maplibre-ios` on Slack for iOS development related questions.

**Reply by georgbachmann** [reply]
The number of affected people is still increasing for my crash and I think it's something memory related. I still can't 100% say it's Mapbox that is causing it, but as my app mostly is under memory pressure when it happens, and Mapbox is using the vast majority of memory consumed by my app, I still think it's related...
As I said, most (like 90% or so) of my users are on very very low memory (according to Crashlytics) when the crash happens.
I notice that when closing my app, Mapbox seems to release a LOT of memory (this seems to be the `_rendererFrontend->reduceMemoryUse();` call). And when then scrolling around in my map, it then reallocates that memory very very quickly. Like 200-400 MB when simply scrolling around in my map very fast. (My map is a combination of raster data and vector data and quite feature rich... like a LOT of layers)
I also noticed that throwing LowMemory Notifications manually reduces memory usage quickly.
So besides that... Is there a way to reduce the amount of cached data that Mapbox uses? Cause when a devices is on such low memory and the scrolling around will allocate that much memory so fast, if can cause problems I guess...

Would be interesting to know other ideas I could try to reduce memory consumption by Mapbox in my app.

**Comment by ovivoda** [comment]
@georgbachmann I suggest to use a memory profiler to spot where the leaks are. Xcode can help for this.

**Reply by georgbachmann** [reply]
Thanks @ovivoda for dropping in here.
Yes, I did that... And all the top allocations are from Mapbox according to Instruments. Also I can see the memory spiking like crazy just by scrolling around on the map. The map that I am using is a private one (as private as such thinks can be ;) ) so I'd rather not share the style-url here, but if you would be interested in helping me out here, I can provide you with a sample project? Would really appreciate it! My C++ skills are very very poor... otherwise I'd love to help myself. But if you want we could do a pair programming, I learn a bit about mapbox and can hopefully then contribute a bit?

**Reply by louwers** [reply]
If you can share it with either me and/or @ovivoda on Slack that would be helpful.

Sounds like a possible space leak which is definitely something we would take seriously and which we would be happy to dig into together.
